velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -140,36 +163,56 @@ class AggregateWindowFunction : public exec::WindowFunction {
     vector_size_t firstRow;
     // Max frame end required for the aggregation.
     vector_size_t lastRow;
-    // Set to indicate if all rows require the same start row.
-    bool fixedFirstRow;
-    // Indicates if the frame end values are non-decreasing. If they are
-    // then we can perform optimizations to do incremental aggregation.
-    bool nonDecreasingFrameEnd;
+    // If all the rows in the block have the same start row, and the
+    // end frame rows are non-decreasing, then the aggregation can be done
+    // incrementally. With incremental aggregation new frame rows are
+    // accumulated over the previous result to obtain the new result.
+    bool incrementalAggregation;
+    // Resume incremental aggregation from the prior block.
+    bool usePreviousAggregate;
   };
 
-  FrameMetadata fillArgVectors(
+  FrameMetadata analyzeFrameValues(
       const vector_size_t* rawFrameStarts,
       const vector_size_t* rawFrameEnds,
       vector_size_t numRows) {
-    // Argument values are needed to compute the aggregate. The aggregate
-    // computation for each row needs the values from its frameStart to frameEnd
-    // indices. So for each apply() we read arguments from the least
-    // frameStart to the highest frameEnd value in the row block.
     vector_size_t firstRow = rawFrameStarts[0];
+    vector_size_t fixedFrameStartRow = firstRow;
     vector_size_t lastRow = rawFrameEnds[0];
 
-    bool nonDecreasingFrameEnd = true;
-    bool fixedFirstRow = true;
+    bool incrementalAggregation = true;
     for (int i = 1; i < numRows; i++) {
       firstRow = std::min(firstRow, rawFrameStarts[i]);
       lastRow = std::max(lastRow, rawFrameEnds[i]);
-      fixedFirstRow &= rawFrameStarts[i] == firstRow;
-      nonDecreasingFrameEnd &= rawFrameEnds[i] >= rawFrameEnds[i - 1];
+      // Incremental aggregation can be done if :
+      // i) All rows have the same frameStart value.
+      // ii) The frame end values are non-decreasing.
+      incrementalAggregation &= (rawFrameStarts[i] == fixedFrameStartRow);
+      incrementalAggregation &= rawFrameEnds[i] >= rawFrameEnds[i - 1];
     }
 
+    bool usePreviousAggregate = false;
+    if (previousFrameMetadata_.has_value()) {
+      auto previousFrameMetadataValue = previousFrameMetadata_.value();
+      // Incremental aggregation continues between blocks if :
+      // i) Their starting firstRow values are the same.
+      // ii) The nonDecreasing frameEnd property is also applicable between the
+      // lastRow of the second and first blocks.
+      if (incrementalAggregation &&
+          previousFrameMetadataValue.incrementalAggregation &&
+          previousFrameMetadataValue.firstRow == firstRow &&
+          previousFrameMetadataValue.lastRow <= lastRow) {
```

Comment:
I'm confused by this check as I'd expect previous.lastRow <= firstRow. Would you clarify?